### ISSUE TITLE
Getting  "Cannot connect Mac <IP> with user mac-4mormwlvqv to jenkins with JNLP"

### ISSUE THREAD

**Issue by OlivierdcMV** [jenkinsci/mac-plugin#20]
### Version report

Jenkins and plugins versions report:

```
Jenkins: 2.303.2
OS: Linux - 5.4.95-42.163.amzn2.x86_64
---
ace-editor:1.1
ansible-tower:0.16.0
apache-httpcomponents-client-4-api:4.5.13-1.0
authentication-tokens:1.4
aws-java-sdk:1.12.89-292.v2712528e879c
aws-java-sdk-cloudformation:1.12.89-292.v2712528e879c
aws-java-sdk-codebuild:1.12.89-292.v2712528e879c
aws-java-sdk-ec2:1.12.89-292.v2712528e879c
aws-java-sdk-ecr:1.12.89-292.v2712528e879c
aws-java-sdk-ecs:1.12.89-292.v2712528e879c
aws-java-sdk-elasticbeanstalk:1.12.89-292.v2712528e879c
aws-java-sdk-iam:1.12.89-292.v2712528e879c
aws-java-sdk-logs:1.12.89-292.v2712528e879c
aws-java-sdk-minimal:1.12.89-292.v2712528e879c
aws-java-sdk-ssm:1.12.89-292.v2712528e879c
aws-lambda:0.5.10
aws-secrets-manager-credentials-provider:0.5.6
aws-secrets-manager-secret-source:0.0.1
bootstrap4-api:4.6.0-3
bootstrap5-api:5.1.1-1
bouncycastle-api:2.25
branch-api:2.7.0
caffeine-api:2.9.2-29.v717aac953ff3
checks-api:1.7.2
cloudbees-disk-usage-simple:0.10
cloudbees-folder:6.16
command-launcher:1.6
configuration-as-code:1.54
copyartifact:1.46.2
credentials:2.6.1
credentials-binding:1.27
dark-theme:0.0.12
dashboard-view:2.18
display-url-api:2.3.5
docker-commons:1.17
docker-workflow:1.26
durable-task:1.39
echarts-api:5.2.1-2
extended-read-permission:3.2
font-awesome-api:5.15.4-1
generic-webhook-trigger:1.77
git:4.10.0
git-client:3.10.0
git-server:1.10
github:1.34.1
github-api:1.133
github-branch-source:2.11.3
github-scm-trait-notification-context:1.1
handlebars:3.0.8
jackson2-api:2.13.0-230.v59243c64b0a5
javadoc:1.6
jaxb:2.3.0
jdk-tool:1.0
jira:3.6
jjwt-api:0.11.2-9.c8b45b8bb173
job-dsl:1.78.1
jquery3-api:3.6.0-2
jsch:0.1.55.2
junit:1.53
kubernetes:1.30.5
kubernetes-client-api:5.4.1
kubernetes-credentials:0.9.0
ldap:2.7
lockable-resources:2.12
mac:1.3.1
mailer:1.34
matrix-auth:2.6.8
matrix-project:1.19
maven-plugin:3.15
metrics:4.0.2.8
modernstatus:1.3
momentjs:1.1.1
okhttp-api:3.14.9-20211029
pipeline-build-step:2.15
pipeline-graph-analysis:1.11
pipeline-input-step:2.12
pipeline-milestone-step:1.3.2
pipeline-model-api:1.9.2
pipeline-model-definition:1.9.2
pipeline-model-extensions:1.9.2
pipeline-rest-api:2.19
pipeline-stage-step:2.5
pipeline-stage-tags-metadata:1.9.2
pipeline-stage-view:2.19
plain-credentials:1.7
plugin-util-api:2.5.0
popper-api:1.16.1-2
popper2-api:2.10.2-1
prometheus:2.0.10
resource-disposer:0.16
rich-text-publisher-plugin:1.4
s3:0.12.0
scm-api:2.6.5
script-security:1.78
simple-theme-plugin:0.7
slack:2.48
snakeyaml-api:1.29.1
ssh-credentials:1.19
sshd:3.1.0
structs:1.23
theme-manager:0.6
timestamper:1.13
token-macro:267.vcdaea6462991
trilead-api:1.0.13
variant:1.4
workflow-aggregator:2.6
workflow-api:2.47
workflow-basic-steps:2.24
workflow-cps:2.94
workflow-cps-global-lib:2.21
workflow-durable-task-step:2.40
workflow-job:2.42
workflow-multibranch:2.26
workflow-scm-step:2.13
workflow-step-api:2.24
workflow-support:3.8
ws-cleanup:0.39
```

- What Operating System are you using (both controller, and any agents involved in the problem)?

```
Jenkins on Kubernetes, Kubernetes cluster build on AMZ Linux -> Controller
OSX Monterey -> Agent
```

### Reproduction steps

- Install plugin in controller
- Configure JNLP port on controller
- Check connectivity from agent to controller
- Configure Mac as described in guide.
- Start job with label associated to the agent config
- Start of agent fails with message in description.

### Results

Expected result:

Agent connects and starts the job.

Actual result:

Error: Getting  "Cannot connect Mac <IP> with user mac-4mormwlvqv to jenkins with JNLP"
We verified connectivity from the agent to the controller and we can see the directories for the users being created.
The logs however show that provisioning was succesfully: <Mac Machine> provisioning successfully completed. We have now 3 computer(s)
Then: Failed to execute command java -jar remoting.jar -jnlpUrl https://jenkinsplay.internal.mobilevikings.be/computer/mac-yzyhk8kewl/slave-agent.jnlp -secret <secret>

**Comment by MagicianTM** [comment]
Hello.
I got the same problem and it is now solved. Actually, I missed the step to allow SSH access to all users (https://plugins.jenkins.io/mac/#enable-ssh-for-all-users). Once I applied this setup, it worked.
Hope it is the same for you

**Comment by OlivierdcMV** [comment]
Hi @MagicianTM ,

Thank you so much for the swift response. We've been breaking our heads on this for a couple of days now and were so excited to get an answer. Sadly we went through the guide multiple times and cross-checked everything we could think of. The access should be ok:

![image](https://user-images.githubusercontent.com/57902049/140048691-d3144a35-77be-43c5-a4be-39d41d9ec898.png)

**Comment by MagicianTM** [comment]
Sorry that it couldn't help you.

If I come back to the problem I've add, I could put the finger on the issue by trying to SSH with the temp Mac user created for the agent (`mac-4mormwlvqv` in you case above). With the logger configured as explained in the doc, the password for that user is visible.
Maybe try this (NB: in the agent properties for your mac cloud, increase the idle minutes so that you can make some trials before the user gets deleted).

For info, in my plugin list, I have `ssh-slaves` (now known as SSH Build Agents plugin). I don't know if it is really used (I just started working with Jenkins), but as you have issue related to SSHing, who knows .... ;)

**Comment by nickjmv** [comment]
Hi @MagicianTM,

I'm a colleague of @OlivierdcMV, we are seeing the following error in the logger:
```
Nov 12, 2021 7:29:12 AM hudson.remoting.jnlp.Main$CuiListener status
INFO: Protocol JNLP4-connect encountered an unexpected exception
java.util.concurrent.ExecutionException: javax.net.ssl.SSLHandshakeException: Public key of the first certificate in chain (subject: C=US, OU=jenkins.io, O=instances, CN=d6dbf8afbf3358e0793a1dadb9e8bdc9) is not in the list of trusted keys
	at org.jenkinsci.remoting.util.SettableFuture.get(SettableFuture.java:223)
	at hudson.remoting.Engine.innerRun(Engine.java:778)
	at hudson.remoting.Engine.run(Engine.java:540)
Caused by: javax.net.ssl.SSLHandshakeException: Public key of the first certificate in chain (subject: C=US, OU=jenkins.io, O=instances, CN=d6dbf8afbf3358e0793a1dadb9e8bdc9) is not in the list of trusted keys
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:131)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:324)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:267)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:262)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1331)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.onConsumeCertificate(CertificateMessage.java:1206)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.consume(CertificateMessage.java:1153)
	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:396)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:444)
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1260)
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1247)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:691)
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:1192)
	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.processRead(SSLEngineFilterLayer.java:380)
	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.onRecv(SSLEngineFilterLayer.java:117)
	at org.jenkinsci.remoting.protocol.ProtocolStack$Ptr.onRecv(ProtocolStack.java:677)
	at org.jenkinsci.remoting.protocol.impl.AckFilterLayer.onRecv(AckFilterLayer.java:255)
	at org.jenkinsci.remoting.protocol.ProtocolStack$Ptr.onRecv(ProtocolStack.java:677)
	at org.jenkinsci.remoting.protocol.NetworkLayer.onRead(NetworkLayer.java:136)
	at org.jenkinsci.remoting.protocol.impl.BIONetworkLayer.access$2200(BIONetworkLayer.java:49)
	at org.jenkinsci.remoting.protocol.impl.BIONetworkLayer$Reader.run(BIONetworkLayer.java:291)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:122)
	at java.base/java.lang.Thread.run(Thread.java:830)
Caused by: java.security.cert.CertificateException: Public key of the first certificate in chain (subject: C=US, OU=jenkins.io, O=instances, CN=d6dbf8afbf3358e0793a1dadb9e8bdc9) is not in the list of trusted keys
	at org.jenkinsci.remoting.protocol.cert.PublicKeyMatchingX509ExtendedTrustManager.checkPublicKey(PublicKeyMatchingX509ExtendedTrustManager.java:219)
	at org.jenkinsci.remoting.protocol.cert.PublicKeyMatchingX509ExtendedTrustManager.checkServerTrusted(PublicKeyMatchingX509ExtendedTrustManager.java:265)
	at org.jenkinsci.remoting.protocol.cert.DelegatingX509ExtendedTrustManager.checkServerTrusted(DelegatingX509ExtendedTrustManager.java:146)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1309)
	... 20 more
```

A quick google search does not give us any more information about this error. Is this anything you have ever faced?

**Comment by MagicianTM** [comment]
Hello @nickjmv ,

I didn't have faced this issue (in my test config, everything was local, and no SSL was set up). But looking into it, my first guess is that you run your Jenkins controller with a self-signed certificate that your agent is not aware of (or at least, it doesn't trust).

So, if possible, try to make your Jenkins working on normal HTTP and configure that address in your Mac cloud (I know it is not the most secure. But make it temporary just to debug your issue).
Else, configure the SSL certificate of your controller with a Let's Encrypt certificate.
Third possibility, update the Java's truststore of your agent: add into it the CA that signed the certificate of your Jenkins controller.

**Comment by nickjmv** [comment]
Hi @MagicianTM,

We tried setting our controller to HTTP which did not resolve the problem. The remoting jar which is downloaded on the agent contains the following content:
```
mac-nrjq8wgdtd@Apps-Mac-mini ~ % cat remoting.jar
<html>
<head><title>308 Permanent Redirect</title></head>
<body>
<center><h1>308 Permanent Redirect</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

Regarding your option 2, we are running the controller on Kubernetes which automatically requests a Let's Encrypt certificate for the Nginx ingress. But Jenkins doesn't seem to use this certificate when talking to the mac agent.
We can't seem to find an option in the helm values file to pass the certificate to the controller.

**Comment by mat1e** [comment]
Hi,

I never faced this, but it looks like this issue : https://issues.jenkins.io/browse/JENKINS-60188

**Comment by Cloudehard** [comment]
Can you test if your Jenkins is sending the complete chain with this command openssl

`openssl s_client    -servername jenkinsplay.internal.mobilevikings.be   -connect jenkinsplay.internal.mobilevikings.be:443`

In your execption you have the message:
`Caused by: java.security.cert.CertificateException: Public key of the first certificate in chain (subject: C=US, OU=jenkins.io, O=instances, CN=d6dbf8afbf3358e0793a1dadb9e8bdc9) is not in the list of trusted keys`
**C=US, OU=jenkins.io, O=instances, CN=d6dbf8afbf3358e0793a1dadb9e8bdc9**

So for me the problem is about the SSL chain you provide

If you doesn't send the only complete chain from your jenkins, the agent cannot started because by default the cacerts from the JRE include only the rootca from AWS, and not the intermediary 
![image](https://user-images.githubusercontent.com/15806177/143718919-cd5d257f-533a-4b56-937a-6bea0bd56be0.png)


`keytool -list  -keystore ./lib/security/cacerts | grep amazon`
Avertissement : utiliser l'option -cacerts pour accéder au fichier de clés cacerts
Entrez le mot de passe du fichier de clés :

*****************  WARNING WARNING WARNING  *****************
* L'intégrité des informations stockées dans votre fichier de clés  *
* n'a PAS été vérifiée. Pour cela, *
* vous devez fournir le mot de passe de votre fichier de clés.                  *
*****************  WARNING WARNING WARNING  *****************

amazonrootca1 [jdk], 25 mai 2015, trustedCertEntry,
amazonrootca2 [jdk], 25 mai 2015, trustedCertEntry,
amazonrootca3 [jdk], 25 mai 2015, trustedCertEntry,
amazonrootca4 [jdk], 25 mai 2015, trustedCertEntry,



The curl command use this files for validate the certificate from your Jenkins:
`/etc/ssl/cert.pem`

**Comment by mat1e** [comment]
@OlivierdcMV, @nickjmv, 

Did you find a solution ? If yes can you publish it for the others users of the plugin ?
This issue concern a configuration problem rather than a bug of the plugin.
If it is ok for you I will close this issue.

**Comment by OlivierdcMV** [comment]
Hi,

After much, much frustration we figured it out.
We run Jenkins in Kubernetes and that is where we made the error.
Using NGINX for our ingress we relied on SNI.

We had problems getting the plugin to work on our main Jenkins node, so we started testing on a separate Jenkins deploy.
Totally forgetting that the JNLP connection does not take SNI into account, we set it up with the test Jenkins with nearly the same values. Hence when the agent tried to create the connection, the response came back from our primary Jenkins as this one is presented first in the NGINX config.

Other than that it was quite some work to get this done in Kubernetes, but I suppose that is out of the scope of your project.
The actual plugin does what it promisses.

Kr,
Olivier
